File: backend/src/models/questionnaire_response.ts
import {
  ModelType,
  prop,
  staticMethod,
  Typegoose,
  instanceMethod,
} from "typegoose";
import Settlement from "./settlement";
import { Question, Section } from "./interfaces";

class QuestionnaireResponse extends Typegoose {
  @staticMethod
  public static findByQuestionnaire(
    this: ModelType<QuestionnaireResponse> & typeof QuestionnaireResponse,
    questionnaireId: string,
  ) {
    return this.find({ questionnaireId });
  }
  @staticMethod
  public static findBySettlement(
    this: ModelType<QuestionnaireResponse> & typeof QuestionnaireResponse,
    settlementId: string,
  ) {
    return this.find({ 'settlements.settlementId': settlementId });
  }
  @staticMethod
  public static findByRespondent(
    this: ModelType<QuestionnaireResponse> & typeof QuestionnaireResponse,
    respondentId: string,
  ) {
    return this.find({ respondentId });
  }
  @staticMethod
  public static findByDate(
    this: ModelType<QuestionnaireResponse> & typeof QuestionnaireResponse,
    date: string,
  ) {
    return this.find({ created: { $gte: date } });
  }
  //
  @prop({ required: true, default: [] })
  public sections!: Section[];
  //
  @prop({ required: true, index: true })
  public userId?: string;
  //
  @prop({ required: true, index: true })
  public questionnaireId?: string;

  @prop({ index: true })
  public respondentId?: string;
  //
  @prop({ index: true })
  public questionnaireResponseId?: string;
  //
  @prop({ required: true, default: [] })
  public settlements!: any[];
  //
  @prop({ required: true, default: new Date().toISOString() })
  public created?: string;
  //
}

export default QuestionnaireResponse;

File: monitorlibrary/lib/data/questionnaire_response.dart
import 'package:meta/meta.dart';
import 'package:monitorlibrary/data/question.dart';
import 'package:monitorlibrary/data/section.dart';

import 'content.dart';

class QuestionnaireResponse {
  String? questionnaireId, userId, respondentId, questionnaireResponseId;
  List<Section>? sections;

  QuestionnaireResponse({required this.questionnaireId, required this.userId,
    required this.respondentId,
    required this.questionnaireResponseId, required this.sections});

  QuestionnaireResponse.fromJson(Map data) {
    this.userId = data['userId'];
    this.questionnaireResponseId = data['questionnaireResponseId'];
    this.sections = data['sections'];
    this.respondentId = data['respondentId'];
    this.questionnaireId = data['questionnaireId'];
  }
  Map<String, dynamic> toJson() {
    List mSecs = [];
    if (sections != null) {
      sections!.forEach((s) {
        mSecs.add(s.toJson());
      });
    }

    Map<String, dynamic> map = {
      'userId': userId,
      'questionnaireResponseId': questionnaireResponseId,
      'sections': mSecs,
      'respondentId': respondentId,
      'questionnaireId': questionnaireId,
    };
    return map;
  }
}
